libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationTelemetry.java

```diff
@@ -0,0 +1,423 @@
+    updateLocationEngine(locationEngine);
+
+    validateAccessToken(accessToken);
+    MapboxNavigationOptions options = navigation.options();
+
+    // Set sdkIdentifier based on if from UI or not
+    String sdkIdentifier = MAPBOX_NAVIGATION_SDK_IDENTIFIER;
+    if (options.isFromNavigationUi()) {
+      sdkIdentifier = MAPBOX_NAVIGATION_UI_SDK_IDENTIFIER;
+    }
+    // Enable extra logging in debug mode
+    MapboxTelemetry.getInstance().setDebugLoggingEnabled(options.isDebugLoggingEnabled());
+
+    String userAgent = String.format("%s/%s", sdkIdentifier, BuildConfig.MAPBOX_NAVIGATION_VERSION_NAME);
+    MapboxTelemetry.getInstance().initialize(context, accessToken, userAgent, sdkIdentifier,
+      BuildConfig.MAPBOX_NAVIGATION_VERSION_NAME);
+    MapboxTelemetry.getInstance().newUserAgent(userAgent);
+
+    // Get the current vendorId
+    vendorId = obtainVendorId(context);
+
+    NavigationMetricsWrapper.sdkIdentifier = sdkIdentifier;
+    NavigationMetricsWrapper.turnstileEvent();
+    // TODO This should be removed when we figure out a solution in NavigationTelemetry
+    // Force pushing a TYPE_MAP_LOAD event to ensure that the Nav turnstile event is sent
+    MapboxTelemetry.getInstance().pushEvent(MapboxEvent.buildMapLoadEvent());
+  }
+
+  /**
+   * Called when navigation is starting for the first time.
+   * Initializes the {@link SessionState}.
+   *
+   * @param directionsRoute first route passed to navigation
+   */
+  void startSession(DirectionsRoute directionsRoute) {
+    navigationSessionState = SessionState.builder()
+      .originalDirectionRoute(directionsRoute)
+      .currentDirectionRoute(directionsRoute)
+      .sessionIdentifier(TelemetryUtils.buildUUID())
+      .routeDistanceCompleted(0)
+      .startTimestamp(new Date())
+      .mockLocation(metricLocation.getLocation().getProvider().equals(MOCK_PROVIDER))
+      .rerouteCount(0)
+      .build();
+  }
+
+  /**
+   * Flushes any remaining events from the reroute / feedback queue and fires
+   * a cancel event indicating a terminated session.
+   */
+  void endSession() {
+    flushEventQueues();
+    NavigationMetricsWrapper.cancelEvent(navigationSessionState, metricProgress,
+      metricLocation.getLocation(), locationEngineName);
+  }
+
+  /**
+   * Called when a new {@link DirectionsRoute} is given in
+   * {@link MapboxNavigation#startNavigation(DirectionsRoute)}.
+   * <p>
+   * At this point, navigation has already begun and the {@link SessionState}
+   * needs to be updated.
+   *
+   * @param directionsRoute new route passed to {@link MapboxNavigation}
+   */
+  void updateSessionRoute(DirectionsRoute directionsRoute) {
+    SessionState.Builder navigationBuilder = navigationSessionState.toBuilder();
+    navigationBuilder.currentDirectionRoute(directionsRoute);
+
+    if (isOffRoute) {
+      // If we are off-route, update the reroute count
+      navigationBuilder.rerouteCount(navigationSessionState.rerouteCount() + 1);
```

Comment:
This should work the way it needs to.  This is only called after the route is updated after an off-route (and we are in an off-route state).  At the time of an off-route event, the reroute count will not be updated (it will capture the current reroute count).